File: frontend/src/services/glossing.ts
import api from '../lib/api';

export interface Morpheme {
  word_index: number;
  morpheme_index: number;
  surface_form: string;
  gloss: string;
  morpheme_type?: string;
  lemma?: string;
}

export interface WordGloss {
  word: string;
  morphemes: Morpheme[];
  full_gloss: string;
}

export interface InterlinearLine {
  original: string[];
  morphemes: string[];
  glosses: string[];
  translation?: string;
}

export interface GlossedText {
  id: string;
  title?: string;
  original_text: string;
  language: string;
  translation?: string;
  difficulty: number;
}

export interface FullGlossedText {
  text: GlossedText;
  lines: InterlinearLine[];
}

export const glossingService = {
  glossText: async (text: string, language: string = 'ru') => {
    const response = await api.post<WordGloss[]>('/glossing/gloss', null, {
      params: { text, language },
    });
    return response.data;
  },

  getInterlinear: async (text: string, language: string = 'ru') => {
    const response = await api.get<InterlinearLine>('/glossing/interlinear', {
      params: { text, language },
    });
    return response.data;
  },

  getText: async (textId: string) => {
    const response = await api.get<FullGlossedText>(`/glossing/texts/${textId}`);
    return response.data;
  },

  listTexts: async (language: string = 'ru') => {
    const response = await api.get<GlossedText[]>('/glossing/texts', {
      params: { language },
    });
    return response.data;
  },
};

File: backend/models/glossing.py
from datetime import datetime
from uuid import uuid4
from sqlalchemy import Column, String, DateTime, Text, ForeignKey, Integer, JSON
from sqlalchemy.orm import relationship

from core.database import Base, GUID


class GlossedText(Base):
    """Text with interlinear glossing annotations"""
    __tablename__ = "glossed_texts"

    id = Column(GUID, primary_key=True, default=uuid4)
    title = Column(String(255))
    original_text = Column(Text, nullable=False)
    language = Column(String(10), nullable=False, default="ru")
    translation = Column(Text)
    source = Column(String(255))
    difficulty = Column(Integer, default=1)
    extra_data = Column(JSON, default=dict)
    created_at = Column(DateTime, default=datetime.utcnow)

    morphemes = relationship("Morpheme", back_populates="text", cascade="all, delete-orphan")


class Morpheme(Base):
    """Individual morpheme with Leipzig glossing"""
    __tablename__ = "morphemes"

    id = Column(GUID, primary_key=True, default=uuid4)
    text_id = Column(GUID, ForeignKey("glossed_texts.id", ondelete="CASCADE"), nullable=False)
    word_index = Column(Integer, nullable=False)  # Position in sentence
    morpheme_index = Column(Integer, nullable=False)  # Position within word
    surface_form = Column(String(255), nullable=False)  # The actual morpheme
    gloss = Column(String(255), nullable=False)  # Leipzig gloss (e.g., "1SG", "ACC", "PRS")
    morpheme_type = Column(String(50))  # root, prefix, suffix, infix, etc.
    lemma = Column(String(255))  # Base form if applicable
    features = Column(JSON, default=dict)  # Additional morphological features
    
    text = relationship("GlossedText", back_populates="morphemes")

